experimental/dds/tree2/src/shared-tree-core/editManager.ts

```diff
@@ -243,59 +246,50 @@ export class EditManager<
-				// no existing branches that are based off of any of the commits being removed.
-				(newTrunkTail as Mutable<GraphCommit<TChangeset>>).parent = this.trunkBase;
-
-				const sequenceNumber =
-					this.trunkMetadata.get(newTrunkTail.revision)?.sequenceNumber ??
-					minimumPossibleSequenceNumber;
-				this.sequenceMap.forRange(
-					minimumPossibleSequenceNumber,
-					sequenceNumber,
-					false,
-					(_seq, { revision }) => {
-						this.trunkMetadata.delete(revision);
-						this.localBranchUndoRedoManager.untrackCommitType(revision);
-					},
-				);
-
-				deleted = this.sequenceMap.deleteRange(
-					minimumPossibleSequenceNumber,
-					sequenceNumber,
-					false,
-				);
+		const [sequenceNumber, latestEvicted] = this.getClosestTrunkCommit(trunkTailSearchKey);
+		// Don't do any work if the commit found by the search is already the tail of the trunk
+		if (latestEvicted !== this.trunkBase) {
+			// The minimum sequence number informs us that all peer branches are at least caught up to the tail commit,
+			// so rebase them accordingly. This is necessary to prevent peer branches from referencing any evicted commits.
+			for (const [, branch] of this.peerLocalBranches) {
+				branch.rebaseOnto(this.trunk, latestEvicted);
 			}
-		} else {
-			// If no trunk commit is found, it means that all trunk commits are below the search key, so evict them all
-			assert(
-				this.trunkBranches.isEmpty ||
-					// This handles the case when there are branches but they are already based off of the origin commit.
-					// TODO:#4918: Investigate if we can handle this case more gracefully by including the origin commit in `sequenceMap`
-					(this.trunkBranches.size === 1 &&
-						this.trunkBranches.minKey() === minimumPossibleSequenceNumber),
-				0x711 /* Expected no outstanding branches when clearing trunk */,
+
+			// This mutation is a performance hack. If commits are truly immutable, then changing the trunk's tail requires
+			// regenerating the entire commit graph. Instead, we can simply chop off the tail like this if we're certain
+			// that there are no outstanding references to any of the commits being removed (other than the references via
+			// the trunk). The peer branches have been rebased to the head of the trunk, the local branch is already rebased
+			// to the head of the trunk, and all other branches are tracked by `trunkBranches` and known to be ahead of or at
+			// `newTrunkBase`. Therefore, no branches should have unique references to any of the commits being evicted here.
+			// We mutate the most recent of the evicted commits to become the new trunk base. That way, any other commits that
+			// have parent pointers to the latest evicted commit will stay linked, even though that it is no longer part of the trunk.
+			const newTrunkBase = latestEvicted as Mutable<GraphCommit<TChangeset>>;
+			newTrunkBase.revision = this.trunkBase.revision;
+			newTrunkBase.change = this.trunkBase.change;
```

Comment:
Is it necessary for correctness to replace the revision and change of this commit? Or is it just for perf (because nothing should need those anymore)?

Reply:
Makes sense. I think explaining as much in comments above lines 267 and 268 would be nice.

Reply:
Good idea, I added some more explanatory comments.